File: scribe/library.py
"""Markdown knowledge library writer.

Layout under the configured library root::

    knowledge/
    ├── topics/<slug>.md                  # one file per topic, appended over time
    ├── meetings/<YYYY-MM-DD>-<slug>.md   # full note per processed meeting
    ├── transcripts/<transcript_id>.txt   # raw transcript with speakers
    └── index.md                          # regenerated on every run

Topic files grow by appending a dated section per meeting; meeting notes and
raw transcripts are written once; the index is rebuilt from scratch each run.
"""

from __future__ import annotations

import re
import unicodedata
from datetime import datetime, timezone
from pathlib import Path

from .analyzer import AnalysisResult, Topic
from .transcriber import TranscriptResult


def slugify(text: str) -> str:
    """Lowercase kebab-case slug, accents stripped (ASCII only)."""
    normalized = unicodedata.normalize("NFKD", text)
    ascii_text = normalized.encode("ascii", "ignore").decode("ascii")
    ascii_text = ascii_text.lower()
    ascii_text = re.sub(r"[^a-z0-9]+", "-", ascii_text)
    return ascii_text.strip("-") or "untitled"


def _fmt_timestamp(ms: int) -> str:
    total_seconds = int(ms) // 1000
    minutes, seconds = divmod(total_seconds, 60)
    hours, minutes = divmod(minutes, 60)
    if hours:
        return f"{hours}:{minutes:02d}:{seconds:02d}"
    return f"{minutes:02d}:{seconds:02d}"


class Library:
    def __init__(self, root: Path) -> None:
        self.root = Path(root)
        self.topics_dir = self.root / "topics"
        self.meetings_dir = self.root / "meetings"
        self.transcripts_dir = self.root / "transcripts"
        for d in (self.topics_dir, self.meetings_dir, self.transcripts_dir):
            d.mkdir(parents=True, exist_ok=True)

    # ------------------------------------------------------------------ read

    def existing_topics(self) -> list[dict]:
        """List of {"slug", "name"} for every topic file, used by the analyzer."""
        topics = []
        for path in sorted(self.topics_dir.glob("*.md")):
            name = path.stem
            try:
                first_line = path.read_text(encoding="utf-8").splitlines()[0]
                if first_line.startswith("# "):
                    name = first_line[2:].strip()
            except (OSError, IndexError):
                pass
            topics.append({"slug": path.stem, "name": name})
        return topics

    # ----------------------------------------------------------------- write

    def write_transcript(self, transcript: TranscriptResult) -> Path:
        """Raw transcript with speaker labels, one utterance per line."""
        path = self.transcripts_dir / f"{transcript.id}.txt"
        lines = []
        if transcript.utterances:
            lines = [f"Speaker {u.speaker}: {u.text}" for u in transcript.utterances]
        elif transcript.text:
            lines = [transcript.text]
        path.write_text("\n\n".join(lines) + "\n", encoding="utf-8")
        return path

    def write_meeting_note(
        self,
        transcript: TranscriptResult,
        analysis: AnalysisResult,
        video_path: Path,
        date: str,
    ) -> Path:
        """Full meeting note; returns the path of the created file."""
        meeting_slug = slugify(analysis.title)
        path = self.meetings_dir / f"{date}-{meeting_slug}.md"

        lines = [
            f"# {analysis.title}",
            "",
            f"- **Date:** {date}",
            f"- **Source video:** `{video_path.name}`",
            f"- **Language:** {transcript.language}",
            f"- **Raw transcript:** [transcript](../transcripts/{transcript.id}.txt)",
            "",
            "## Summary",
            "",
            analysis.summary or "(no summary)",
            "",
        ]

        if analysis.key_points:
            lines += ["## Key points", ""]
            lines += [f"- {point}" for point in analysis.key_points]
            lines += [""]

        if analysis.decisions:
            lines += ["## Decisions", ""]
            lines += [f"- {decision}" for decision in analysis.decisions]
            lines += [""]

        if analysis.action_items:
            lines += ["## Action items", ""]
            lines += [
                f"- [ ] **{item['owner']}**: {item['task']}" for item in analysis.action_items
            ]
            lines += [""]

        if transcript.chapters:
            lines += ["## Chapters", ""]
            for ch in transcript.chapters:
                start = _fmt_timestamp(ch.start)
                end = _fmt_timestamp(ch.end)
                label = ch.headline or ch.gist or "Chapter"
                lines.append(f"- `{start}–{end}` {label}")
            lines += [""]

        path.write_text("\n".join(lines), encoding="utf-8")
        return path

    def append_topic_section(
        self, topic: Topic, date: str, meeting_title: str, meeting_file: Path
    ) -> Path:
        """Append a dated section to a topic file, creating it if needed."""
        path = self.topics_dir / f"{topic.slug}.md"
        if not path.exists():
            path.write_text(f"# {topic.name}\n", encoding="utf-8")

        link = f"../meetings/{meeting_file.name}"
        section = (
            f"\n## {date} — {meeting_title}\n\n"
            f"{topic.content.strip()}\n\n"
            f"*Source: [{meeting_title}]({link})*\n"
        )
        with open(path, "a", encoding="utf-8") as fh:
            fh.write(section)
        return path

    def rebuild_index(self) -> Path:
        """Regenerate index.md: topics with last update + meetings, newest first."""
        topics = []
        for path in sorted(self.topics_dir.glob("*.md")):
            name = path.stem
            last_update = ""
            try:
                text = path.read_text(encoding="utf-8")
                first_line = text.splitlines()[0]
                if first_line.startswith("# "):
                    name = first_line[2:].strip()
                dates = re.findall(r"^## (\d{4}-\d{2}-\d{2})", text, re.MULTILINE)
                if dates:
                    last_update = max(dates)
            except (OSError, IndexError):
                pass
            topics.append((path.stem, name, last_update))

        meetings = sorted(self.meetings_dir.glob("*.md"), reverse=True)

        lines = ["# Knowledge library", ""]
        lines += ["## Topics", ""]
        if topics:
            for slug, name, last_update in topics:
                suffix = f" — last updated {last_update}" if last_update else ""
                lines.append(f"- [{name}](topics/{slug}.md){suffix}")
        else:
            lines.append("(no topics yet)")
        lines += ["", "## Meetings", ""]
        if meetings:
            for path in meetings:
                try:
                    first_line = path.read_text(encoding="utf-8").splitlines()[0]
                    title = first_line[2:].strip() if first_line.startswith("# ") else path.stem
                except (OSError, IndexError):
                    title = path.stem
                lines.append(f"- [{path.stem}](meetings/{path.name}) — {title}")
        else:
            lines.append("(no meetings yet)")
        lines.append("")

        index_path = self.root / "index.md"
        index_path.write_text("\n".join(lines), encoding="utf-8")
        return index_path

    # ------------------------------------------------------------- high level

    def add_meeting(
        self, transcript: TranscriptResult, analysis: AnalysisResult, video_path: Path
    ) -> Path:
        """Persist everything for one processed meeting; returns the note path."""
        date = datetime.now(timezone.utc).strftime("%Y-%m-%d")
        self.write_transcript(transcript)
        note_path = self.write_meeting_note(transcript, analysis, video_path, date)
        for topic in analysis.topics:
            self.append_topic_section(topic, date, analysis.title, note_path)
        self.rebuild_index()
        return note_path

File: internal/library/library.go
// Package library writes the Markdown knowledge library.
//
// Layout under the configured library root:
//
//	knowledge/
//	├── topics/<slug>.md                  # one file per topic, appended over time
//	├── meetings/<YYYY-MM-DD>-<slug>.md   # full note per processed meeting
//	├── transcripts/<transcript_id>.txt   # raw transcript with speakers
//	└── index.md                          # regenerated on every run
//
// Topic files grow by appending a dated section per meeting; meeting notes
// and raw transcripts are written once; the index is rebuilt from scratch
// each run. The generated Markdown is byte-for-byte identical to the legacy
// Python writer (scribe/library.py).
package library

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"strings"
	"time"
	"unicode"

	"golang.org/x/text/runes"
	"golang.org/x/text/transform"
	"golang.org/x/text/unicode/norm"

	"github.com/fernando143/patro/internal/types"
)

var (
	// slugTransformer mirrors Python's NFKD + encode("ascii", "ignore"):
	// decompose, drop combining marks, then drop any remaining non-ASCII.
	slugTransformer = transform.Chain(
		norm.NFKD,
		runes.Remove(runes.In(unicode.Mn)),
		runes.Map(func(r rune) rune {
			if r > unicode.MaxASCII {
				return -1
			}
			return r
		}),
	)
	nonSlugChars  = regexp.MustCompile(`[^a-z0-9]+`)
	sectionDateRe = regexp.MustCompile(`(?m)^## (\d{4}-\d{2}-\d{2})`)
)

// Slugify converts text to a lowercase kebab-case slug, accents stripped
// (ASCII only). An empty result becomes "untitled".
func Slugify(s string) string {
	ascii, _, _ := transform.String(slugTransformer, s)
	slug := nonSlugChars.ReplaceAllString(strings.ToLower(ascii), "-")
	if slug = strings.Trim(slug, "-"); slug != "" {
		return slug
	}
	return "untitled"
}

// formatTimestamp renders milliseconds as MM:SS, or H:MM:SS once the
// duration reaches one hour.
func formatTimestamp(ms int64) string {
	totalSeconds := ms / 1000
	seconds := totalSeconds % 60
	minutes := (totalSeconds / 60) % 60
	hours := totalSeconds / 3600
	if hours > 0 {
		return fmt.Sprintf("%d:%02d:%02d", hours, minutes, seconds)
	}
	return fmt.Sprintf("%02d:%02d", minutes, seconds)
}

// Library writes transcripts, meeting notes, topic files and the index
// under a root directory.
type Library struct {
	Root           string
	TopicsDir      string
	MeetingsDir    string
	TranscriptsDir string
}

// NewLibrary creates the library directory layout under root and returns
// the Library handle.
func NewLibrary(root string) (*Library, error) {
	l := &Library{
		Root:           root,
		TopicsDir:      filepath.Join(root, "topics"),
		MeetingsDir:    filepath.Join(root, "meetings"),
		TranscriptsDir: filepath.Join(root, "transcripts"),
	}
	for _, d := range []string{l.TopicsDir, l.MeetingsDir, l.TranscriptsDir} {
		if err := os.MkdirAll(d, 0o755); err != nil {
			return nil, err
		}
	}
	return l, nil
}

// ExistingTopics lists a {slug, name} reference for every topic file, used
// by the analyzer.
func (l *Library) ExistingTopics() []types.TopicRef {
	files, err := filepath.Glob(filepath.Join(l.TopicsDir, "*.md"))
	if err != nil {
		return nil
	}
	topics := make([]types.TopicRef, 0, len(files))
	for _, path := range files {
		slug := stem(path)
		name, _ := topicInfo(path, slug)
		topics = append(topics, types.TopicRef{Slug: slug, Name: name})
	}
	return topics
}

// WriteTranscript writes the raw transcript with speaker labels, one
// utterance per paragraph, and returns its path.
func (l *Library) WriteTranscript(t *types.TranscriptResult) (string, error) {
	path := filepath.Join(l.TranscriptsDir, t.ID+".txt")
	var lines []string
	if len(t.Utterances) > 0 {
		lines = make([]string, 0, len(t.Utterances))
		for _, u := range t.Utterances {
			lines = append(lines, "Speaker "+u.Speaker+": "+u.Text)
		}
	} else if t.Text != "" {
		lines = []string{t.Text}
	}
	if err := os.WriteFile(path, []byte(strings.Join(lines, "\n\n")+"\n"), 0o644); err != nil {
		return "", err
	}
	return path, nil
}

// WriteMeetingNote writes the full meeting note and returns the path of
// the created file.
func (l *Library) WriteMeetingNote(t *types.TranscriptResult, a *types.AnalysisResult, videoPath, date string) (string, error) {
	path := filepath.Join(l.MeetingsDir, date+"-"+Slugify(a.Title)+".md")

	summary := a.Summary
	if summary == "" {
		summary = "(no summary)"
	}

	lines := []string{
		"# " + a.Title,
		"",
		"- **Date:** " + date,
		"- **Source video:** `" + filepath.Base(videoPath) + "`",
		"- **Language:** " + t.Language,
		"- **Raw transcript:** [transcript](../transcripts/" + t.ID + ".txt)",
		"",
		"## Summary",
		"",
		summary,
		"",
	}

	if len(a.KeyPoints) > 0 {
		lines = append(lines, "## Key points", "")
		for _, point := range a.KeyPoints {
			lines = append(lines, "- "+point)
		}
		lines = append(lines, "")
	}

	if len(a.Decisions) > 0 {
		lines = append(lines, "## Decisions", "")
		for _, decision := range a.Decisions {
			lines = append(lines, "- "+decision)
		}
		lines = append(lines, "")
	}

	if len(a.ActionItems) > 0 {
		lines = append(lines, "## Action items", "")
		for _, item := range a.ActionItems {
			lines = append(lines, fmt.Sprintf("- [ ] **%s**: %s", item.Owner, item.Task))
		}
		lines = append(lines, "")
	}

	if len(t.Chapters) > 0 {
		lines = append(lines, "## Chapters", "")
		for _, ch := range t.Chapters {
			label := ch.Headline
			if label == "" {
				label = ch.Gist
			}
			if label == "" {
				label = "Chapter"
			}
			lines = append(lines, fmt.Sprintf("- `%s–%s` %s", formatTimestamp(ch.Start), formatTimestamp(ch.End), label))
		}
		lines = append(lines, "")
	}

	if err := os.WriteFile(path, []byte(strings.Join(lines, "\n")), 0o644); err != nil {
		return "", err
	}
	return path, nil
}

// AppendTopicSection appends a dated section to a topic file, creating the
// file with a "# <name>" heading when needed. meetingFile is the path of
// the meeting note; only its base name is linked. Returns the topic path.
func (l *Library) AppendTopicSection(topic types.Topic, date, meetingTitle, meetingFile string) (string, error) {
	path := filepath.Join(l.TopicsDir, topic.Slug+".md")
	if _, err := os.Stat(path); os.IsNotExist(err) {
		if err := os.WriteFile(path, []byte("# "+topic.Name+"\n"), 0o644); err != nil {
			return "", err
		}
	}

	link := "../meetings/" + filepath.Base(meetingFile)
	section := fmt.Sprintf("\n## %s — %s\n\n%s\n\n*Source: [%s](%s)*\n",
		date, meetingTitle, strings.TrimSpace(topic.Content), meetingTitle, link)

	f, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		return "", err
	}
	defer f.Close()
	if _, err := f.WriteString(section); err != nil {
		return "", err
	}
	return path, nil
}

// RebuildIndex regenerates index.md: topics with their last update date
// and meetings, newest first. Returns the index path.
func (l *Library) RebuildIndex() (string, error) {
	type topicEntry struct {
		slug, name, lastUpdate string
	}

	topicFiles, err := filepath.Glob(filepath.Join(l.TopicsDir, "*.md"))
	if err != nil {
		return "", err
	}
	topics := make([]topicEntry, 0, len(topicFiles))
	for _, path := range topicFiles {
		slug := stem(path)
		name, lastUpdate := topicInfo(path, slug)
		topics = append(topics, topicEntry{slug, name, lastUpdate})
	}

	meetings, err := filepath.Glob(filepath.Join(l.MeetingsDir, "*.md"))
	if err != nil {
		return "", err
	}
	slices.Reverse(meetings)

	lines := []string{"# Knowledge library", "", "## Topics", ""}
	if len(topics) > 0 {
		for _, t := range topics {
			suffix := ""
			if t.lastUpdate != "" {
				suffix = " — last updated " + t.lastUpdate
			}
			lines = append(lines, fmt.Sprintf("- [%s](topics/%s.md)%s", t.name, t.slug, suffix))
		}
	} else {
		lines = append(lines, "(no topics yet)")
	}
	lines = append(lines, "", "## Meetings", "")
	if len(meetings) > 0 {
		for _, path := range meetings {
			stemName := stem(path)
			title := stemName
			if data, err := os.ReadFile(path); err == nil {
				if line, ok := firstLine(string(data)); ok && strings.HasPrefix(line, "# ") {
					title = strings.TrimSpace(line[2:])
				}
			}
			lines = append(lines, fmt.Sprintf("- [%s](meetings/%s) — %s", stemName, filepath.Base(path), title))
		}
	} else {
		lines = append(lines, "(no meetings yet)")
	}
	lines = append(lines, "")

	indexPath := filepath.Join(l.Root, "index.md")
	if err := os.WriteFile(indexPath, []byte(strings.Join(lines, "\n")), 0o644); err != nil {
		return "", err
	}
	return indexPath, nil
}

// AddMeeting persists everything for one processed meeting and returns the
// note path.
func (l *Library) AddMeeting(t *types.TranscriptResult, a *types.AnalysisResult, videoPath string) (string, error) {
	date := time.Now().UTC().Format("2006-01-02")
	if _, err := l.WriteTranscript(t); err != nil {
		return "", err
	}
	notePath, err := l.WriteMeetingNote(t, a, videoPath, date)
	if err != nil {
		return "", err
	}
	for _, topic := range a.Topics {
		if _, err := l.AppendTopicSection(topic, date, a.Title, notePath); err != nil {
			return "", err
		}
	}
	if _, err := l.RebuildIndex(); err != nil {
		return "", err
	}
	return notePath, nil
}

// stem returns the file name without its final extension (Python's
// Path.stem).
func stem(path string) string {
	base := filepath.Base(path)
	return strings.TrimSuffix(base, filepath.Ext(base))
}

// firstLine returns the first line of text, or false when text is empty.
func firstLine(text string) (string, bool) {
	if text == "" {
		return "", false
	}
	line := text
	if i := strings.IndexByte(line, '\n'); i >= 0 {
		line = line[:i]
	}
	return strings.TrimSuffix(line, "\r"), true
}

// topicInfo extracts a topic file's display name (first "# " heading,
// falling back to the slug) and the most recent "## <date>" section date.
// Unreadable or empty files yield the slug and no date.
func topicInfo(path, slug string) (name, lastUpdate string) {
	name = slug
	data, err := os.ReadFile(path)
	if err != nil {
		return name, ""
	}
	text := string(data)
	line, ok := firstLine(text)
	if !ok {
		return name, ""
	}
	if strings.HasPrefix(line, "# ") {
		name = strings.TrimSpace(line[2:])
	}
	for _, m := range sectionDateRe.FindAllStringSubmatch(text, -1) {
		if m[1] > lastUpdate {
			lastUpdate = m[1]
		}
	}
	return name, lastUpdate
}
